cnd/src/db/integration_tests.rs

```diff
@@ -1,56 +1,116 @@
 
 macro_rules! db_roundtrip_test {
-    ($alpha_ledger:ident, $beta_ledger:ident, $alpha_asset:ident, $beta_asset:ident) => {
+    ($alpha_ledger:ident, $beta_ledger:ident, $alpha_asset:ident, $beta_asset:ident, $expected_swap_types_fn:expr) => {
         paste::item! {
             #[test]
-            #[allow(non_snake_case)]
+            #[allow(non_snake_case, clippy::redundant_closure_call)]
             fn [<roundtrip_test_ $alpha_ledger _ $beta_ledger _ $alpha_asset _ $beta_asset>]() {
-                fn prop(swap_id: Quickcheck<SwapId>,
-                    request: Quickcheck<Request<$alpha_ledger, $beta_ledger, $alpha_asset, $beta_asset>>,
-                    accept: Quickcheck<Accept<$alpha_ledger, $beta_ledger>>) -> anyhow::Result<bool> {
-                let db_path = tempfile::Builder::new()
-                        .prefix(&swap_id.to_string())
-                        .suffix(".sqlite")
-                        .tempfile()?
-                        .into_temp_path();
-
-                let db = Sqlite::new(&db_path)?;
-
-                let saved_request = Request {
-                    swap_id: *swap_id,
-                    ..*request
-                };
-                let saved_accept = Accept {
-                    swap_id: *swap_id,
-                    ..*accept
-                };
-
-                db.save_message(saved_request.clone())?;
-                db.save_message(saved_accept.clone())?;
-
-                let (loaded_request, loaded_accept) = db.load_accepted_swap(*swap_id)?;
-
-                Ok(saved_request == loaded_request && saved_accept == loaded_accept)
-            }
+                fn prop(swap: Quickcheck<Swap>,
+                        request: Quickcheck<Request<$alpha_ledger, $beta_ledger, $alpha_asset, $beta_asset>>,
+                        accept: Quickcheck<Accept<$alpha_ledger, $beta_ledger>>,
+                ) -> anyhow::Result<bool> {
+
+                    // unpack the swap from the generic newtype
+                    let Swap { swap_id, role, counterparty } = swap.0;
+
+                    // construct the expected swap types from the function we get passed in order to enrich it with the role
+                    let expected_swap_types = ($expected_swap_types_fn)(role);
+
+                    let db = Sqlite::new(&Path::new(":memory:"))?;
+
+                    let saved_swap = Swap {
+                        swap_id,
+                        role,
+                        counterparty
+                    };
+                    let saved_request = Request {
+                        swap_id,
+                        ..*request
+                    };
+                    let saved_accept = Accept {
+                        swap_id,
+                        ..*accept
+                    };
+
+                    let (loaded_swap, loaded_request, loaded_accept, loaded_swap_types) =
+                    async_std::task::block_on::<_, Result<_, anyhow::Error>>(async {
```

Comment:
I understand you spawn a new runtime for that right?
Wouldn't it be better to not block here but return a future so that it's all awaited by a runtime spawned in the `main()`?

Reply:
This is an integration test :)
We are basically the "main"!

Reply:
lol indeed.